nginx/1.25.3-zlib-ng.patch

```diff
@@ -0,0 +1,332 @@
+      */
+ 
+-    if (!ngx_http_gzip_assume_zlib_ng) {
+-        ctx->allocated = 8192 + 16 + (1 << (wbits + 2))
+-                         + (1 << (memlevel + 9));
++    if (!ngx_http_gzip_assume_intel) {
++        ctx->allocated = 8192 + (1 << (wbits + 2)) + (1 << (memlevel + 9));
+ 
+     } else {
+         /*
+-         * Another zlib variant, https://github.com/zlib-ng/zlib-ng.
+-         * It used to force window bits to 13 for fast compression level,
+-         * uses (64 + sizeof(void*)) additional space on all allocations
+-         * for alignment, 16-byte padding in one of window-sized buffers,
+-         * and 128K hash.
++         * A zlib variant from Intel, https://github.com/jtkukunas/zlib.
++         * It can force window bits to 13 for fast compression level,
++         * on processors with SSE 4.2 it uses 64K hash instead of scaling
++         * it from the specified memory level, and also introduces
++         * 16-byte padding in one out of the two window-sized buffers.
+          */
+ 
+         if (conf->level == 1) {
+             wbits = ngx_max(wbits, 13);
+         }
+ 
+-        ctx->allocated = 8192 + 16 + (1 << (wbits + 2))
+-                         + 131072 + (1 << (memlevel + 8))
+-                         + 4 * (64 + sizeof(void*));
+-        ctx->zlib_ng = 1;
++        // zlib-ng
++        ctx->allocated = 8192                                /* deflate_state + padding + extra? */
++                         + (1 << (wbits + 1)) + 8            /* s->window + s->prev + padding */
++                         + (1 << (17)) + 8                   /* s->head */
++                         + (1 << (ngx_max(memlevel, 8) + 8)) /* s->pending_buf */
++                         + (1 << (memlevel + 8));            /* not sure */
++
++        ctx->intel = 1;
+     }
+ }
+ 
+@@ -621,7 +627,7 @@
+     ctx->zstream.zfree = ngx_http_gzip_filter_free;
+     ctx->zstream.opaque = ctx;
+ 
+-    rc = deflateInit2(&ctx->zstream, (int) conf->level, Z_DEFLATED,
++    rc = ZPREFIX(deflateInit2)(&ctx->zstream, (int) conf->level, Z_DEFLATED,
+                       ctx->wbits + 16, ctx->memlevel, Z_DEFAULT_STRATEGY);
+ 
+     if (rc != Z_OK) {
+@@ -756,7 +762,7 @@
+                  ctx->zstream.avail_in, ctx->zstream.avail_out,
+                  ctx->flush, ctx->redo);
+ 
+-    rc = deflate(&ctx->zstream, ctx->flush);
++    rc = ZPREFIX(deflate)(&ctx->zstream, ctx->flush);
+ 
+     if (rc != Z_OK && rc != Z_STREAM_END && rc != Z_BUF_ERROR) {
+         ngx_log_error(NGX_LOG_ALERT, r->connection->log, 0,
+@@ -880,7 +886,7 @@
+     ctx->zin = ctx->zstream.total_in;
+     ctx->zout = ctx->zstream.total_out;
+ 
+-    rc = deflateEnd(&ctx->zstream);
++    rc = ZPREFIX(deflateEnd)(&ctx->zstream);
+ 
+     if (rc != Z_OK) {
+         ngx_log_error(NGX_LOG_ALERT, r->connection->log, 0,
+@@ -929,16 +935,13 @@
+ 
+     alloc = items * size;
+ 
+-    if (items == 1 && alloc % 512 != 0 && alloc < 8192
```

Comment:
Why is this part of the code changed? Did it expose a bug with zlib-ng?
This might increase memory allocations, and that would have to be compensated with the memory calculations above.
This should behave the same both for zlib-ng-compat and zlib-ng-native.